PYIC-7580: Handle access-denied error response

AmritSidhu opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Rather than a VC we may receive an asynchronous error response off the queue. 
In the case of a user abandon/abort (where the user has opted to abandon the app) the error response will contain an access-denied 
At present when we process an asychronous response error, we simply update the CRI response item ('pending record') with status error regardless of the specific error code. 
We should use a new status type abandon for this scenario and update the CRI response accordingly (with error used for all other error codes).

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-7580](https://govukverify.atlassian.net/browse/PYIC-7580)

## Checklists

### Environment variables or secrets

<!-- Delete if changes DO include new environment variables or secrets -->
- [x] No environment variables or secrets were added or changed


[PYIC-7580]: https://govukverify.atlassian.net/browse/PYIC-7580?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

AmritSidhu committed:
PYIC-7580: Handle access-denied error response

Rather than a VC we may receive an asynchronous error response off the queue. In the case of a user abandon/abort (where the user has opted to abandon the app) the error response will contain an access-denied
At present when we process an asychronous response error, we simply update the CRI response item ('pending record') with status error regardless of the specific error code.
We should use a new status type abandon for this scenario and update the CRI response accordingly (with error used for all other error codes).

DanCorderIPV commented on lambdas/process-async-cri-credential/src/test/java/uk/gov/di/ipv/core/processasynccricredential/ProcessAsyncCriCredentialHandlerTest.java:
```diff
@@ -176,7 +178,27 @@ void shouldProcessSuccessfullyIfEvcsFailsAndEvcsReadsIsNotEnabled() throws Excep
+        verify(auditService, times(1)).sendAuditEvent(auditEventCaptor.capture());
+        List<AuditEvent> auditEvents = auditEventCaptor.getAllValues();
+        assertEquals(1, auditEvents.size());
+        assertEquals(AuditEventTypes.IPV_F2F_CRI_VC_ERROR, auditEvents.get(0).getEventName());
```
Maybe check the response item status in this test too?

DanCorderIPV commented on lambdas/process-async-cri-credential/src/main/java/uk/gov/di/ipv/core/processasynccricredential/ProcessAsyncCriCredentialHandler.java:
```diff
@@ -157,7 +157,11 @@ private void processErrorAsyncCriResponse(ErrorAsyncCriResponse errorAsyncCriRes
         criResponseItem.ifPresent(
                 responseItem -> {
-                    responseItem.setStatus(CriResponseService.STATUS_ERROR);
+                    if ("access_denied".equals(errorAsyncCriResponse.getError())) {
```
`access_denied` should probably be a constant here

DanCorderIPV commented on lambdas/process-async-cri-credential/src/main/java/uk/gov/di/ipv/core/processasynccricredential/ProcessAsyncCriCredentialHandler.java:
```diff
@@ -157,7 +157,11 @@ private void processErrorAsyncCriResponse(ErrorAsyncCriResponse errorAsyncCriRes
-                    responseItem.setStatus(CriResponseService.STATUS_ERROR);
+                    if ("access_denied".equals(errorAsyncCriResponse.getError())) {
+                        responseItem.setStatus(CriResponseService.STATUS_ABANDON);
+                    } else {
```
It would be good to have a way of noticing any new response error codes.
Either by logging an error here, or even throwing an exception if the code isn't one we recognise.

AmritSidhu replied:
As part of this ticket it says we should use a new status type abandon specifically access_denied. 'error' to be used for all other error codes. So I dont think as part of this ticket we want to start differentiating or treating other error codes any differently yet

AmritSidhu committed:
PYIC-7580: Use constant for access_denied error

Update exisiting test to check for STATUS_ERROR to show the difference with the new STATUS_ABANDON
